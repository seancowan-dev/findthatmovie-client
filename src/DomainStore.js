import {observable,action,toJS} from 'mobx';
import uuid from "uuid";

class DomainStore{
    // App Primary 
     @observable updateType = "";
     @observable updateFolderId = "";
     @observable updateNoteId = "";
     @observable currentNote = null;
     @observable done = "";
     @observable updated = null;
     @observable errorMsg = "";
     @observable notes = {};
     @observable folders = {};

    // Update Component
     @observable folderNotes = [];
     @observable selectedFolder = "";
     @observable selectedFolderId = "";
     @observable selectedFolderIdOld = "";
     @observable oldNoteId = null;
     @observable currentNoteId = null;
     @observable migrateChecked = false;
     @observable note = {
         id: null,
         name: null,
         folderid: null,
         modified: null,
         content: null
     };

    // Local methods to use with fetch api methods
     handleErrors(response) { // prepares error message for HTTP request errors
      if (response.ok === true) {
          return response.json();
      } else {
          throw new Error("Code " + response.status + " Message: " + response.statusText)
      }
    }
    hex(string) {
      let source = unescape(encodeURIComponent(string));
      let hex = '';
        for (var i = 0; i < source.length; i++) {
            hex += source.charCodeAt(i).toString(16)
        };
        return hex;
    }

  // Fetch API Methods

  @action async getFolders() {
    return await fetch(`http://localhost:8000/api/folders`, {
      method: 'GET',
      headers: {
        'content-type': 'application/json'
      },
    })
    .then(response => this.handleErrors(response))
    .then(resJSON => { return resJSON })
    .catch(e => alert(e));
  }

  @action async getNotes() {
    return await fetch(`http://localhost:8000/api/notes`, {
      method: 'GET',
      headers: {
        'content-type': 'application/json'
      },
    })
    .then(response => this.handleErrors(response))
    .then(resJSON => { return resJSON })
    .catch(e => alert(e));
  }

  @action async deleteNote(id) {
    return await fetch(`http://localhost:8000/api/notes/${id}`, {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json'
      },
    })
    .then(response => { 
      if (response.ok === true) {
        this.promises();
    } else {
        throw new Error("Code " + response.status + " Message: " + response.statusText)
    }
    })
    .catch(e => alert(e));

  }

  @action async addNote(event) {
    event.preventDefault();
    let workingArr = Array.from(event.target.parentNode.childNodes);
    let date = new Date(Date.now());
    let note = {
      id: uuid.v4(),
      name: "",
      modified: date.toISOString(),
      folderid: "",
      content: ""
    };
    workingArr.filter(child => {
      if(child.name === "add-note-content") {
        note.content = child.value;
      }
      if (child.name === "add-note-title") {
        note.name = child.value;
      }
      if (child.name === "select-note-folder") {
        note.folderid = child.selectedOptions[0].id;
      }
      return null
    });
    await fetch(`http://localhost:8000/api/notes/`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(note)
    })
    .then(response => {
      if (response.ok === true) {
        this.promises()
    } else {
        throw new Error("Code " + response.status + " Message: " + response.statusText)
    }
    }).catch(e => alert(e));    
  }

  @action async addFolder(event) {
    event.preventDefault();
    let workingArr = Array.from(event.target.parentNode.childNodes);
    let folder = {
      id: "",
      name: "",
    };
    workingArr.filter(child => {
      if(child.name === "add-folder-title") {
        folder.name = child.value;
        folder.id = uuid.v4();
      }
      return null
    });

    await fetch(`http://localhost:8000/api/folders/`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(folder)
    })
    .then(response => {
      if (response.ok === true) {
        this.promises()
    } else {
        throw new Error("Code " + response.status + " Message: " + response.statusText)
    }
    }).catch(e => alert(e));
  }

  @action async deleteFolder(id) {
    await fetch(`http://localhost:8000/api/folders/${id}`, {
      method: `DELETE`,
      headers: {
        'content-type': 'application/json'
      },

    })
    .then(response => { 
      if (response.ok === true) {
        this.promises()
    } else {
        throw new Error("Code " + response.status + " Message: " + response.statusText)
    }
    })
    .catch(e => alert(e));
  }

  @action async updateFolder(id, updates) {
    await fetch(`http://localhost:8000/api/folders/${id}`, {
      method: `PATCH`,
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(updates),
    })
    .then(response => {
      if (response.ok === true) {
        this.promises()
    } else {
        throw new Error("Code " + response.status + " Message: " + response.statusText)
    }
    })
    .catch(e => alert(e));
  }

  @action async updateNote(id, updates) {
    await fetch(`http://localhost:8000/api/notes/${id}`, {
      method: `PATCH`,
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(updates),
    })
    .then(response => {
        if (response.ok === true) {
          this.promises()
        } else {
          throw new Error("Code " + response.status + " Message: " + response.statusText)
        }
    })
    .catch(e => alert(e));
  }

  @action promises() {
    Promise.all([this.getFolders(), this.getNotes()]).then(res => {
      this.folders = res[0];
      this.notes = res[1];
      this.done = true;
    });
  }
}
export default new DomainStore();