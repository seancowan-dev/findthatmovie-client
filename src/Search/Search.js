import React from 'react';
import { Button, FormControl, InputLabel, TextField, Select, Grid, MenuItem } from '@material-ui/core';
// FormControlLabel, FormHelperText, Checkbox, 
import { observer, inject } from 'mobx-react';
import uuid from 'uuid';
import { navigate } from 'hookrouter';

const Search = inject("searchStore", "dataStore", "helpers")(observer((props) => {

    let years = [];
    let genres = [];
    let countries = [];
    let ratings = [];

        for (let i = 0; i < props.dataStore.genres.length; i++) {
            genres[i] = <MenuItem key={uuid.v4()} value={props.dataStore.genres[i].code}>{props.dataStore.genres[i].genre}</MenuItem>
        }
        for (let i = 0; i < props.dataStore.years.length; i++) {
            years[i] = <MenuItem key={uuid.v4()} value={props.dataStore.years[i].code}>{props.dataStore.years[i].year}</MenuItem>
        }
        for (let i = 0; i < props.dataStore.countries.length; i++) {
            countries[i] = <MenuItem key={uuid.v4()} value={props.dataStore.countries[i].value}>{props.dataStore.countries[i].country}</MenuItem>
        }
        for (let i = 0; i < 11; i++) {
            if (i === 0) {
                ratings[i] = <MenuItem key={uuid.v4()} value={"Min"}>Min</MenuItem>
            } else {
                ratings[i] = <MenuItem key={uuid.v4()} value={i}>{i}</MenuItem>
            }
        }
    return <form className="search-form">
    <Grid 
        container
        direction="column"
        alignItems="center"
    >
    <Grid item xs={12}>
        <FormControl className="user-query-control">
            <TextField
                id="user-query-input"
                className="user-query-input"
                label="Enter keyword(s) or title"
                value={props.searchStore.searchParams.userQuery}
                onChange={(e) => {
                    props.searchStore.searchParams.userQuery = e.target.value;
                }}
            />
        </FormControl>
        <FormControl className="select-year-control">
            <InputLabel id="select-year-label">Year</InputLabel>
            <Select
                labelId="select-year-label"
                id="select-year"
                className="select-year"
                value={props.searchStore.searchParams.selectedYear}
                onChange={(e) => {
                    props.searchStore.searchParams.selectedYear = e.target.value;
                }}
                >
                {years}
            </Select>
        </FormControl>
        </Grid>
        <Grid item xs={12}>
        <FormControl className="select-genre-control">
            <InputLabel id="select-genre-label">Genre</InputLabel>
            <Select
                labelId="select-genre-label"
                id="select-genre"
                className="select-genre"
                value={props.searchStore.searchParams.selectedGenre}
                onChange={(e) => {
                    props.searchStore.searchParams.selectedGenre = e.target.value;
                }}
                >
                {genres}
            </Select>
        </FormControl>
        {/* |Temporarily disabled because this requires a combined API call which will be written on Node later |
        <FormControl className="select-country-control">
            <InputLabel id="select-country-label">Country</InputLabel>
            <Select
                labelId="select-country-label"
                id="select-country"
                className="select-country"
                value={props.searchStore.searchParams.selectedCountry}
                onChange={(e) => {
                    props.searchStore.searchParams.selectedCountry = e.target.value;
                }}
                >
                {countries}
            </Select>
        </FormControl> */}
        <FormControl className="select-rating-control-min">
            <InputLabel id="select-rating-label-min">Rating</InputLabel>
            <Select
                labelId="select-rating-label-min"
                id="select-rating-min"
                className="select-rating-min"
                value={props.searchStore.searchParams.selectedRatingMin}
                onChange={(e) => {
                    props.searchStore.searchParams.selectedRatingMin = e.target.value;
                }}
                >
                {ratings}
            </Select>
        </FormControl>
        </Grid>
        <Button
        variant="contained"
        onClick={(e) => {
            navigate('/results'); // Move user forward to next results page | Breadcrumb this later|

            props.searchStore.getMovieList(false).then(res => {
                props.searchStore.searchResults = []; // Make sure old results are removed
                if (res.results !== undefined || res.results.length !== 0) {
                    props.searchStore.total_pages = res.total_pages;
                    props.searchStore.searchResults.push(res);
                    props.searchStore.searchState = "display";
                }
                if (res.results === undefined || res.results.length === 0) {
                    props.searchStore.searchState = "none";
                }
            });
        }}
        >
            Search
        </Button>
    </Grid>

</form>
  }));

export default Search;
