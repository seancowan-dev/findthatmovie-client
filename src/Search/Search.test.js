import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from "mobx-react";
import DomainStore from '../DomainStore';
import Search from './Search';

const store = {
    dataStore: DomainStore.dataStore,
    searchStore: DomainStore.searchStore,
    helpers: DomainStore.helpers,
    userStore: DomainStore.userStore,
    validators: DomainStore.validators
  }

describe('| Search Test Object |', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Provider {...store}><Search debug /></Provider>);
    expect(component).toMatchSnapshot();
  });
  it('should render correctly in "production" mode', () => {
    const component = shallow(<Provider {...store}><Search production /></Provider>);
    expect(component).toMatchSnapshot();
  });
});