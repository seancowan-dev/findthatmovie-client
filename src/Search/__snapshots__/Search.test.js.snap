// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`| Search Test Object | should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`| Search Test Object | should render correctly in "production" mode 1`] = `ShallowWrapper {}`;
